### PR TITLE
ENH: Add batch, quick mode to annex file_has_content

### PR DESCRIPTION
**Pull request by debanjum**

### CODE REVIEW

File: datalad/support/annexrepo.py [hunk 1]
```diff
@@ -549,44 +552,72 @@ def get_file_key(self, file_):
...
+        objects = []
+        if batch:
+            objects = self._batched.get('find', path=self.path)(files)
+        else:
+            for f in files:
+                try:
+                    obj, er = self._run_annex_command('find', annex_options=[f], expect_fail=True)
+                    objects.append(obj)
+                except CommandError:
+                    objects.append('')
```
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: I guess we could just store None's not just empty strings to make things more explicit and Pythonic

File: datalad/support/annexrepo.py [hunk 2]
```diff
@@ -549,44 +552,72 @@ def get_file_key(self, file_):
...
+
+        return objects
+
     @normalize_paths
-    def file_has_content(self, files):
+    def file_has_content(self, files, allow_quick=True, batch=True):
```
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: by default `batch=False`

File: datalad/support/annexrepo.py [hunk 3]
```diff
@@ -549,44 +552,72 @@ def get_file_key(self, file_):
...
+            for f in files:
+                filepath = opj(self.path, f)
+                target_path = ''
+                if islink(filepath):                    # if symlink
+                    target_path = readlink(filepath)    # find link target
+                    # convert to absolute path if not
+                    target_path = target_path \
+                        if isabs(target_path) \
+                        else opj(dirname(filepath), target_path)
+                out.append(exists(target_path) and islink(filepath))
```
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: what if it is a symlink just pointing to another present file which is not even under annex?  I guess you would need to add a check that `realpath(target_path).startswith(opj(self.path, '.git', 'annex', 'objects')))` ?

File: datalad/support/annexrepo.py [hunk 4]
```diff
@@ -549,44 +552,72 @@ def get_file_key(self, file_):
...
+            out = []
+            for f in files:
+                filepath = opj(self.path, f)
+                target_path = ''
+                if islink(filepath):                    # if symlink
+                    target_path = readlink(filepath)    # find link target
+                    # convert to absolute path if not
+                    target_path = target_path \
+                        if isabs(target_path) \
+                        else opj(dirname(filepath), target_path)
```
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: I would say that
```python
if not isabs(target_path):
  target_path = opj(...)
```
is easier to read etc

File: datalad/support/annexrepo.py (hunk 3 above)
Comment (yarikoptic) [inline datalad/support/annexrepo.py]: and it is not a link (and we are not in direct mode) we can right away store False... why drag some empty `target_path` around etc for that case?

**Comment by yarikoptic** [comment]
interesting... fails on osx... I thought the problem is in outdated annex, upgraded it but still fails.  Any clues @debanjum  or should I look in detail?

**Comment by debanjum** [comment]
no, was trying to figure it out myself, haven't changed too much b/w this and the previous commit.

**Comment by yarikoptic** [comment]
please merge master and push again -- build with freshier annex (on osx) must be fixed now

**Pull request merged**